Update C++ rules (#3220)
SonarSource/rspec: rules/S5018/cfamily/rule.adoc

== Why is this an issue?

> Move operations (move constructor, move assignment operator) are about efficient resource stealing. When stealing resources from the source, you don't have to allocate any memory or perform any other operation that might fail. This is why most people will expect move operations to be non-throwing.

Reviewer: I'm uneasy with "stealing" as may convey it's a bad thing to do. Maybe "transfer ownership" is better? WDYT?

```suggestion
Move operations (move constructor, move assignment operator) are about efficiently transferring resource ownership. When transfering resources from the source, you don't have to allocate any memory or perform any other operation that might fail. This is why most people will expect move operations to be non-throwing.
```

Feel free to adapt as you like.

Author: Indeed, I agree that "stealing" is not the best word. I like "transferring ownership" better, thanks!

Additionally, if a move operation fails, the source object can have been partially altered by the move, making recovery very tricky or just impossible. Therefore, to ensure robustness, some functions (for instance, ``++std::move_if_noexcept++``, used by ``++std::vector++``) will decide to copy your object if its move operations are not decorated with ``++noexcept++``. This can significantly slow down your program.

If you can not implement your move operations so that they never throw, you may only provide copy operations that will be safer to use.

> Swap operations are similar to move operations in that they should be equivalent to moving two objects into each other. So if you add a swap function to your type, it should be _noexcept_ too.

Reviewer: For consistency:
```suggestion
Swap operations are similar to move operations in that they should be equivalent to moving two objects into each other. So if you add a swap function to your type, it should be `noexcept` too.
```

Note that you should not write your move operations for most classes but rely on the "Rule-of-Zero" (S4963).

> This rule raises an issue when a move or swap operation is not _noexcept_, which can happen in two cases:

Reviewer: For consistency
```suggestion
This rule raises an issue when a move or swap operation is not `noexcept`, which can happen in two cases:
```
